netplan/cli/commands/set.py

```diff
@@ -71,15 +73,55 @@ def command_set(self):
         with tempfile.TemporaryFile() as tmp:
             libnetplan.create_yaml_patch(yaml_path, value, tmp)
             tmp.flush()
+
             tmp.seek(0, io.SEEK_SET)
             parser.load_nullable_fields(tmp)
+            if filename:
+                # Validate the full YAML hierarchy without ignoring any NetDefs.
+                # Exit early if this YAML patch would lead to an invalid config.
+                # XXX: The origin file of each individual YAML setting/stanza
+                #      should be tracked individually, to avoid this
+                #      double-parsing workaround (LP: #2003727)
+                try:
+                    validation_parser = libnetplan.Parser()
+                    tmp.seek(0, io.SEEK_SET)
+                    validation_parser.load_nullable_fields(tmp)
+                    validation_parser.load_yaml_hierarchy(self.root_dir)
+                    tmp.seek(0, io.SEEK_SET)
+                    validation_parser.load_yaml(tmp)
+                    validation_state = libnetplan.State()
+                    validation_state.import_parser_results(validation_parser)
+                except libnetplan.LibNetplanException as e:
+                    logging.error('New settings invalidate existing configuration: ' + str(e))
+                    sys.exit(1)
+
+                # Load nullable overrides, keeping track of netdefs/globals that are
+                # supposed to be overridden/extended by the yaml patch. Those netdefs
+                # and globals must end up in the origin-hint file (i.e. <filename>),
+                # have they been defined in previously parsed YAML file or not.
+                tmp.seek(0, io.SEEK_SET)
+                parser.load_nullable_overrides(tmp, constraint=filename)
+
+            # Parse the full hierarchy.
+            # If output filename was given, ignore any netdef (or global value)
+            # that is supposed to be defined (overriden/extended) inside the
+            # new or updated origin-hint (i.e. <filename> + yaml patch applied).
+            # This way we can avoid updates to certain netdefs/globals to be
+            # redirected into existing YAML files (defining those same stanzas)
+            # or ignored, but have them written out to the single origin-hint
+            # file, using state.write_yaml_file() below.
             parser.load_yaml_hierarchy(self.root_dir)
+
+            # Load YAML patch.
+            # If output filename was given any update from the yaml patch will
+            # be applied inside the output file, ignoring instances of the same
+            # stanzas in the rest of the YAML hierarchy.
             tmp.seek(0, io.SEEK_SET)
             parser.load_yaml(tmp)
 
         state = libnetplan.State()
         state.import_parser_results(parser)
-        if self.origin_hint:
+        if filename:
             state.write_yaml_file(filename, self.root_dir)
```

Comment:
suggestion (not-blocking): keep the main validation in the `state` state, and consider the de-facto incomplete state with the overridden fields as the secondary

This would allow you to only have one branching point: At first you go through the motions of pre-loading the patch, loading the sources, import them in the state, and only once you have a valid state do you actually bother to check whether you have an actual filename target. At that point, you either write out your state, or you create a second one with the additional override state, which you proceed to dump out.

Bonus points:
* that would solve my earlier nitpick about validation errors ;)
* if validation fails we'll have done less work

Reply:
Thanks, this should make the code much easier to grasp. I like that!